File: solutions/haskell/space-age/1/src/SpaceAge.hs
-- for my Shree DR.MDD
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury | Venus | Earth | Mars | Jupiter | Saturn | Uranus | Neptune
              deriving (Eq, Show)

ageOn :: Planet -> Int -> Float
ageOn orb secs = yearsOnEarth / orbitalRatio where
    secsPerEarth = 31557600
    yearsOnEarth = fromIntegral secs / secsPerEarth
    orbitalRatio =
        case orb of 
            Mercury -> 0.2408467
            Venus   -> 0.61519726
            Earth   -> 1.0
            Mars    -> 1.8808158
            Jupiter -> 11.862615
            Saturn  -> 29.447498
            Uranus  -> 84.016846
            Neptune -> 164.79132

File: solutions/elixir/space-age/1/lib/space_age.ex
# For my Shree DR.MDD

defmodule SpaceAge do
  @type world ::
          :mercury
          | :venus
          | :earth
          | :mars
          | :jupiter
          | :saturn
          | :uranus
          | :neptune

  @earth_year_seconds 365.25 * 24 * 60 * 60

  @spec age_on(world, pos_integer) :: float | {:error, String.t()}
  def age_on(world, seconds) do
    compute_years(orbit_time(world), seconds)
  end

  @spec orbit_time(world) :: float
  defp orbit_time(world) do
    case world do
      :earth -> @earth_year_seconds
      :mercury -> 0.2408467 * @earth_year_seconds
      :venus -> 0.61519726 * @earth_year_seconds
      :mars -> 1.8808158 * @earth_year_seconds
      :jupiter -> 11.862615 * @earth_year_seconds
      :saturn -> 29.447498 * @earth_year_seconds
      :uranus -> 84.016846 * @earth_year_seconds
      :neptune -> 164.79132 * @earth_year_seconds
      _ -> 0
    end
  end

  @invalid_planet_msg "not a planet"

  @spec compute_years(float, pos_integer) :: float | {:error, String.t()}
  defp compute_years(0, _), do: {:error, @invalid_planet_msg}
  defp compute_years(period, secs), do: {:ok, secs / period}
end
